src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -276,14 +277,23 @@ private void RemoveProducer(PubSubPublisherState producer)
             State.Producers.Remove(producer);
         }
 
-        public async Task UnregisterConsumer(GuidId subscriptionId, StreamId streamId)
+        public async Task UnregisterConsumer(GuidId subscriptionId, StreamId streamId, bool observerRemoved)
         {
             counterConsumersRemoved.Increment();
             if (State.Consumers.Any(c => c.IsFaulted && c.Equals(subscriptionId)))
                 throw new FaultedSubscriptionException(subscriptionId, streamId);
 
             try
             {
+                if (!observerRemoved)
+                {
+                    //remove observer from consumer extension if not removed yet
+                    //this would happen when subscription removed by calling StreamSubscriptionManager
+                    var tasks = State.Consumers.Where(c => c.Equals(subscriptionId))
```

Comment:
I'll review today.  As for the onSubscriptionRemove, I agree it probably won't get much use, but I do think it's an important signal.  The problem I see with it is that I'm not clear how to wire it up.  I'm ok with leaving it off, but I suggest creating an Issue so we can revisit it.